clang/lib/CConv/ConstraintVariables.cpp

```diff
@@ -493,12 +518,25 @@ PointerVariableConstraint::mkString(EnvironmentMap &E,
                                     bool ForItype,
                                     bool EmitPointee) {
   std::ostringstream Ss;
-  std::ostringstream Pss;
-  unsigned CaratsToAdd = 0;
+  // This deque will store all the type strings that need to pushed
+  // to the end of the type string. This is typically things like
+  // closing delimiters.
+  std::deque<std::string> EndStrs;
+  // This will store stacked array decls to ensure correct order
+  // We encounter constant arrays variables in the reverse order they
+  // need to appear in, so the LIFO structure reverses these annotations
+  std::stack<std::string> CheckedArrs;
+  // Have we emitted the string for the base type
   bool EmittedBase = false;
+  // Have we emitted the name of the variable yet?
   bool EmittedName = false;
-  bool EmittedCheckedAnnotation = false;
-  if (EmitName == false && hasItype() == false)
+  // Was the last cycle an Array?
+  bool PrevArr = false;
+  // Is the entire type so far an array?
+  bool AllArrays = true;
+  // Are we in a sequence of arrays
+  bool ArrayRun = false;
```

Comment:
Rather than three booleans, would it make sense to use an enumerated type to indicate the "state" you are in? If the booleans really are independent (i.e., there are legitimate situations for each true/false combination of every variable) then they should stay as they are.

Reply:
Got rid of one of the booleans, wasn't actually being used, unsure why it wasn't caught in a warning.
Other two need to be independant